### ISSUE TITLE
curl-7.78 doesn't accept arbitary spaces at the end of parsed url anymore

### ISSUE THREAD

**Issue by stefson** [DeaDBeeF-Player/deadbeef#2648]
hey there, all stable version including deadbeef-1.8.8 fail to play this internet stream radio: https://wdr-1live-live.icecastssl.wdr.de/wdr/1live/live/mp3/128/stream.mp3

this is a regression, I listen to it all the time. There was a recent curl update *coughs* 

Good news is, I believe it is fixed in recent git master. 

Deadbeef version: 1.8.4 and 1.8.8
OS: Linux-amd64

I was unable to get your static-devel binaries to work on my system, so can you please:

1. confirm it is working on recent git master
2. can you please help finding and backporting the possible fix? it's your code in the end :D 

the log doesn't say anything usefull: 

```
resume: track -1 pos -1.000000 playlist -1
gtkui plugin compiled for gtk version: 2.24.33

(deadbeef:31519): Gdk-CRITICAL **: 22:26:45.864: IA__gdk_window_get_position: assertion 'GDK_IS_WINDOW (window)' failed

(deadbeef:31519): Gdk-CRITICAL **: 22:26:45.864: IA__gdk_window_get_position: assertion 'GDK_IS_WINDOW (window)' failed

(deadbeef:31519): Gdk-CRITICAL **: 22:26:45.865: gdk_window_invalidate_maybe_recurse_full: assertion 'GDK_IS_WINDOW (window)' failed
connecting button tray signals
Failed to play track: https://wdr-1live-live.icecastssl.wdr.de/wdr/1live/live/mp3/128/stream.mp3 
```

**Comment by stefson** [comment]
possible regression window: https://github.com/DeaDBeeF-Player/deadbeef/compare/1.8...master

**Comment by stefson** [comment]
config of deadbeef-1.8.8:

```
./configure --prefix=/usr --build=x86_64-pc-linux-gnu --host=x86_64-pc-linux-gnu --mandir=/usr/share/man --infodir=/usr/share/info --datadir=/usr/share --sysconfdir=/etc --localstatedir=/var/lib --disable-silent-rules --docdir=/usr/share/doc/deadbeef-1.8.8 --htmldir=/usr/share/doc/deadbeef-1.8.8/html --libdir=/usr/lib64 --disable-coreaudio --disable-portable --disable-static --enable-ffmpeg --enable-aac --disable-adplug --enable-alac --enable-alsa --enable-cdda --disable-cdda-paranoia --enable-converter --enable-artwork --enable-artwork-imlib2 --disable-artwork-network --enable-vfs-curl --disable-dca --disable-dumb --disable-supereq --enable-flac --disable-gme --enable-gtk2 --disable-gtk3 --enable-hotkeys --disable-lfm --disable-notify --enable-src --enable-m3u --disable-ffap --enable-libmad --disable-wildmidi --enable-mms --disable-mono2stereo --disable-libmpg123 --enable-musepack --enable-nls --disable-nullout --disable-opus --disable-oss --disable-pltbrowser --disable-psf --disable-pulse --disable-sc68 --disable-shellexecui --enable-shn --disable-sid --enable-sndfile --enable-tta --enable-vorbis --disable-vtx --enable-wavpack --enable-wma --enable-vfs-zip
```

config of deadbeef-git: 
```
./configure --prefix=/usr --build=x86_64-pc-linux-gnu --host=x86_64-pc-linux-gnu --mandir=/usr/share/man --infodir=/usr/share/info --datadir=/usr/share --sysconfdir=/etc --localstatedir=/var/lib --disable-silent-rules --docdir=/usr/share/doc/deadbeef-9999 --htmldir=/usr/share/doc/deadbeef-9999/html --with-sysroot=/ --libdir=/usr/lib64 --enable-nls --enable-threads --enable-alsa --disable-oss --disable-pulse --enable-gtk2 --disable-gtk3 --enable-vfs-curl --disable-lfm --disable-artwork --disable-sid --enable-libmad --disable-ffap --enable-hotkeys --enable-vorbis --enable-ffmpeg --enable-flac --enable-sndfile --enable-wavpack --enable-cdda --disable-gme --disable-notify --enable-musepack --disable-wildmidi --enable-tta --disable-dca --enable-aac --enable-mms --enable-src --enable-vfs-zip --disable-dumb --disable-shn --enable-alac --enable-wma --disable-opus
```

**Comment by Alexey-Yakovenko** [comment]
It's working in both 1.8.8 and master for me on mac. I'm still trying to test in linux, having issues with VirtualBox.

**Comment by Alexey-Yakovenko** [comment]
In the meantime, could please enable logging in the curl plugin (it can be toggled in the plugin settings), and post the output when trying to play this stream?

**Comment by Alexey-Yakovenko** [comment]
In the static builds on linux it doesn't work, because libcurl is built without https, this is a known problem, even though I can't find a github issue about it. Basically, to get this to work -- you have to build the curl plugin with libcurl which itself is built with HTTPS support, at the very least.

A local linux build using libcurl with https support works fine here.

If you enable curl logging in deadbeef -- you will immediately see the problem right away.

**Comment by stefson** [comment]
here is a log of deadbeef-1.8.4 with the curl logging enabled 

```
resume: track -1 pos -1.000000 playlist -1
http_get_content_type
http_get_content_type waiting for response...
vfs_curl: started loading data https://wdr-1live-live.icecastssl.wdr.de/wdr/1live/live/mp3/128/stream.mp3 
vfs_curl: calling curl_easy_perform (status=0)...
vfs_curl: curl_easy_perform retval=3
curl error:

vfs_curl: break loop
vfs_curl: thread ended normally
http_close 0x7f757c011220
abort vfs_curl stream: 1
added file to abort list: 1
http_close done
Failed to play track: https://wdr-1live-live.icecastssl.wdr.de/wdr/1live/live/mp3/128/stream.mp3 
```

**Comment by Alexey-Yakovenko** [comment]
libcurl returns error code 3 which corresponds to CURLE_URL_MALFORMAT.
This means that there's some other internal issue with libcurl, I don't think it has anything to do with deadbeef.
I suggest to try running curl command line utility with the same url, and see if this works at all.

**Comment by stefson** [comment]
I don't understand why deadbeef-git works, while 1.8.8 doesn't

its possible to fetch the url with curl itself, the dumped mp3 plays fine. 

which flavour of libcurl are you using on ubuntu?

**Comment by Alexey-Yakovenko** [comment]
I diffed 1.8.8 vs master branch -- there's nothing relevant to your problem. Basically, there's nothing to backport.

**Comment by Alexey-Yakovenko** [comment]
```
$ curl --version
curl 7.68.0 (x86_64-pc-linux-gnu) libcurl/7.68.0 OpenSSL/1.1.1f zlib/1.2.11 brotli/1.0.7 libidn2/2.2.0 libpsl/0.21.0 (+libidn2/2.2.0) libssh/0.9.3/openssl/zlib nghttp2/1.40.0 librtmp/2.3
Release-Date: 2020-01-08
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS brotli GSS-API HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets
```

**Comment by stefson** [comment]
is it libcurl4-gnutls or libcurl4-openssl though?

**Comment by Alexey-Yakovenko** [comment]
I don't know how to find this out. Either of the two should work just fine.

**Comment by Alexey-Yakovenko** [comment]
found
```
$ curl-config --ssl-backends
OpenSSL
```

**Comment by Alexey-Yakovenko** [comment]
If you think that the problem is caused by libcurl update -- I suggest trying to downgrade libcurl, and see if this helps.

**Comment by stefson** [comment]
> If you think that the problem is caused by libcurl update -- I suggest trying to downgrade libcurl, and see if this helps.

this was too obvious, but really downgrading to curl-7.77.0 reenabled the stream.

**Comment by stefson** [comment]
really that shouldn't happen for the n-th patch release of such a major package 

maybe there were some symbols dropped or what have you not, I will try to speak with upstream and ask them on their opinion

**Comment by stefson** [comment]
do you use static linking? there is an upstream bug: https://github.com/curl/curl/issues/7475

**Comment by Alexey-Yakovenko** [comment]
libcurl is linked statically in the official builds, but not in the ones you build yourself.
I don't think this is related to the problem you are experiencing.

**Comment by Alexey-Yakovenko** [comment]
especially since the official builds don't use libcurl 7.78..

**Comment by stefson** [comment]
soo, mystery solved. 

Can you please add a safety net to vfs_curl, i.e. not allow users to paste an url with a blank at the end?

**Comment by Alexey-Yakovenko** [comment]
I will consider :)

**Comment by stefson** [comment]
excellent, what is now going to happen if the user parses an url with an arbitary space at the end? Will it give a warning, will it remove the arbitary space?

**Comment by Alexey-Yakovenko** [comment]
It will remove any whitespace / linebreak characters from the text before use. No warnings.

**Comment by stefson** [comment]
I backported to v1.8.8, its working :dancers::dancers: 

can you imagine to give it an official backport to 1.8 branch?

**Comment by Alexey-Yakovenko** [comment]
No, why? I'm not planning to make new 1.8 series releases

**Comment by stefson** [comment]
oops, I just noticed the 1.9 beta is in the pipeline.

maybe I can ask for a backport to that branch instead? :-S

**Comment by Alexey-Yakovenko** [comment]
Yes, all current fixes will be merged to 1.9 branch
